### PR TITLE
Add Base.Process library

### LINKED ISSUE #492: Implement a System Interface Library
### Summary
The user should be able to interact with other functionality of the operating system on which their programs run, so we need to provide a set of functionality that allows them to do so.

### Value
Users can interact with their OS.

### Specification

- [ ] Provide the functionality for running external commands on the system.
- [ ] Provide the functionality for writing to std streams. 
- [ ] Think briefly about what this means and how it should interact with the container in the cloud product. 
- [ ] (Optional) If needed, implement `mask`. 

### Acceptance Criteria & Test Cases

- The above functionality has been implemented.
- The above functionality has been rigorously tested.

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #492 

Add `Base.Process` library with operations for creating and interacting with sub-processes.

- add `System.create_process` primitive to spawn new system processes
- add `Base.Process` library

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: distribution/std-lib/Base/src/Process.enso [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+## Create a process using a builder returning the result of execution.
+
+   > Example
+     Create a script redirecting the input to stdout:
+         builder = Process_Builder "bash" ["-c", "read line; echo -n $line"] "test"
+         Process.create_process builder
+     The result is:
+         Process_Result 0 "test" ""
+Process.create : Process_Builder -> Process_Result
```
Comment (iamrecursion) [inline distribution/std-lib/Base/src/Process.enso:41]: The `Process_Builder` type should be re-exported from the standard library itself, and used from there in here. Not imported directly from builtins.

File: distribution/std-lib/Base/src/Process.enso [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    result.exit_code
+
+## Call a command with a list of arguments.
+
+   > Example
+     Call the `echo` command with arguments
+         Process.shell "echo" ["-n", "Hello!"]
+     The result is printed to stdout:
+         Hello!
+Process.shell : String -> Vector -> Number
```
Comment (iamrecursion) [inline distribution/std-lib/Base/src/Process.enso]: `shell` seems like a strange name for this.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/system/CreateProcessNode.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    System.arraycopy(arguments.getItems(), 0, cmd, 1, (int) arguments.getArraySize());
+    TruffleProcessBuilder pb = ctx.getEnvironment().newProcessBuilder(cmd);
+
+    try {
+      Process p = pb.start();
+      ByteArrayInputStream in = new ByteArrayInputStream(input.getBytes());
+      ByteArrayOutputStream out = new ByteArrayOutputStream();
+      ByteArrayOutputStream err = new ByteArrayOutputStream();
+
+      try (OutputStream processin = p.getOutputStream()) {
```
Comment (iamrecursion) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/system/CreateProcessNode.java]: Please use standard naming conventions: `processIn`.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/system/CreateProcessNode.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+          stdin = in;
+        }
+        int nread;
+        byte[] buf = new byte[8096];
+        while (stdin.available() > 0 && (nread = stdin.read(buf)) != -1) {
+          processin.write(buf, 0, nread);
+        }
+      } catch (IOException ignored) {
+        // Getting the output stream of a finished process results in an IOException.
+        // We can ignore it at this point.
```
Comment (iamrecursion) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/system/CreateProcessNode.java:75]: Is this the _only_ circumstance under which an IOException can occur here?

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/system/CreateProcessNode.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+          stdout = out;
+        }
+        int nread;
+        byte[] buf = new byte[8096];
+        while ((nread = processout.read(buf)) != -1) {
+          stdout.write(buf, 0, nread);
+        }
+      }
+
+      try (InputStream processerr = p.getErrorStream()) {
```
Comment (iamrecursion) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/system/CreateProcessNode.java]: Naming.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/system/CreateProcessNode.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+          processin.write(buf, 0, nread);
+        }
+      } catch (IOException ignored) {
+        // Getting the output stream of a finished process results in an IOException.
+        // We can ignore it at this point.
+      }
+
+      p.waitFor();
+
+      try (InputStream processout = p.getInputStream()) {
```
Comment (iamrecursion) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/system/CreateProcessNode.java]: Naming.

Comment (iamrecursion) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/system/CreateProcessNode.java]: Naming.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/system/CreateProcessNode.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+          stderr = err;
+        }
+        int nread;
+        byte[] buf = new byte[8096];
+        while ((nread = processerr.read(buf)) != -1) {
+          stderr.write(buf, 0, nread);
+        }
+      }
+
+      long exitCode = p.exitValue();
```
Comment (iamrecursion) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/system/CreateProcessNode.java:108]: Is exit codes as longs the abstraction we really want? In a language like Enso we'd rather a sum type `ExitSuccess | ExitFailure value | ...`.

**Review by iamrecursion — CHANGES_REQUESTED** [review]
A great start, but there are some issues with the API and API usage that I'd like to see resolved.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/system/CreateProcessNode.java (hunk 7 above)
Comment (4e6) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/system/CreateProcessNode.java:108]: Fixed in stdlib

File: distribution/std-lib/Base/src/Process.enso [hunk 8]
```diff
@@ -0,0 +1,44 @@
+import Base.Process.Exit_Code
```
Comment (iamrecursion) [inline distribution/std-lib/Base/src/Process.enso:1]: I still have issues with naming in this file. I think maybe the following:

`Process.execute`
`Process.run_command`
`Process.build`

File: distribution/std-lib/Base/src/Process/Exit_Code.enso [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+    type Exit_Failure code
+
+    ## Convert exit code to a number.
+    to_number = case this of
+        Exit_Success -> 0
+        Exit_Failure code -> code
+
+## Create exit code from a number.
+from_number : Number -> Exit_Code
+from_number code = if code == 0 then Exit_Success else Exit_Failure code
```
Comment (iamrecursion) [inline distribution/std-lib/Base/src/Process/Exit_Code.enso]: Needs examples.

File: distribution/std-lib/Base/src/System/Platform.enso [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+## PRIVATE
+   Create an Os object from text.
+from_text: Text -> Os
+from_text os =
+    if os == "linux" then Linux else if os == "macos" then MacOS else if os == "windows" then Windows else Unknown
+
+## Return the type of operating system.
+os : Os
+os = Platform.from_text System.os
```
Comment (iamrecursion) [inline distribution/std-lib/Base/src/System/Platform.enso]: Needs examples.

**Review by iamrecursion — CHANGES_REQUESTED** [review]

File: distribution/std-lib/Base/src/Process.enso (hunk 8 above)
Comment (4e6) [reply distribution/std-lib/Base/src/Process.enso:1]: `execute` and `run_command` are better, agree. But I still like `create` more than `build`.

Comment (iamrecursion) [reply distribution/std-lib/Base/src/Process.enso:1]: Okay, leave it as `create`.

**Review by iamrecursion — APPROVED** [review]

**Pull request merged**